### PR TITLE
Build .rpm packages for linux arm64

### PR DESCRIPTION
**Pull request by wtgodbe**
Part of https://github.com/dotnet/aspnetcore/issues/40253, https://github.com/dotnet/aspnetcore/issues/40252. Test build: https://dev.azure.com/dnceng/internal/_build/results?buildId=1638723&view=results

CC @NikolaMilosavljevic

### CODE REVIEW

File: .azure/pipelines/ci.yml [hunk 1]
```diff
@@ -474,17 +474,27 @@ stages:
...
+      steps:
+      - script: ./eng/build.sh
+            --ci
+            --arch arm64
+            --pack
+            --all
+            --no-build-nodejs
+            --no-build-java
+            -p:OnlyPackPlatformSpecificPackages=true
+            -p:AssetManifestFileName=aspnetcore-Linux_arm64.xml
```
Comment (dougbu) [inline .azure/pipelines/ci.yml:486]: Official build should break if this setting exists in anything except the last step of the job. You may be :ok: w/o `$(_PublishArgs)` but that just leaves this line as redundant and confusing.

**Review by dougbu — CHANGES_REQUESTED** [review]

File: .azure/pipelines/ci.yml (hunk 1 above)
Comment (wtgodbe) [reply .azure/pipelines/ci.yml:486]: Good catch - I copied this from the Linux x64 job, so I guess the duplication exists there too. I'll remove it from both.

**Comment by wtgodbe** [comment]
New build: https://dev.azure.com/dnceng/internal/_build/results?buildId=1638654&view=results

**Comment by wtgodbe** [comment]
https://dev.azure.com/dnceng/internal/_build/results?buildId=1638723&view=results

**Review by dougbu — APPROVED** [review]
Looks fine but I'm not sure why C++ projects are building (and breaking) in the updated build job&hellip;

**Comment by wtgodbe** [comment]
> Looks fine but I'm not sure why C++ projects are building (and breaking) in the updated build job…

I think https://github.com/dotnet/aspnetcore/pull/40485/commits/ca0c6d28edcaa4da4c393ad2b1fefb055a11638e will fix it. https://dev.azure.com/dnceng/internal/_build/results?buildId=1638815&view=results

**Comment by wtgodbe** [comment]
Got this mostly working internally, will push to the public branch once it's fully working (and I'm removed the binlog I added): https://dev.azure.com/dnceng/internal/_build/results?buildId=1638974&view=results

**Comment by wtgodbe** [comment]
Ok, latest internal builds looks like it worked. @NikolaMilosavljevic how did you test the runtime linux-arm64 installers?

**Comment by dougbu** [comment]
@NikolaMilosavljevic something looks wrong here. All of our Linux builds produce files named e.g. aspnetcore-targeting-pack-7.0.0-preview.3.22151.16.rpm but we use platform-specific names for the Windows targeting packs. What is the intention here -- aspnetcore-targeting-pack-7.0.0-preview.3.22151.16.rpm or aspnetcore-targeting-pack-7.0.0-preview.3.22151.16-arm64.rpm❔

We'll need to make a change either way -- stop uploading the extra targeting packs (and remove duplicates from the asset manifests) or change the names.

**Comment by dougbu** [comment]
Also why bother w/ Mariner-specific targeting packs❔ aspnetcore-targeting-pack-7.0.0-preview.3.22151.16-cm.1.rpm and aspnetcore-targeting-pack-7.0.0-preview.3.22151.16-cm.2.rpm look weird if we're not spelling out the architecture. Those two files and aspnetcore-targeting-pack-7.0.0-preview.3.22151.16.rpm will _never_ have different content.

/fyi we produce just one aspnetcore-targeting-pack-7.0.0-preview.3.22151.16.deb file.

**Comment by dougbu** [comment]
If it matters, I don't understand the runtime (really Microsoft.DotNet.SharedFramework.Sdk from Arcade) choices in this area:

image

There's also a dotnet-targeting-pack-7.0.0-preview3-22154-x64.rpm file. Why not just one dotnet-targeting-pack-7.0.0-preview3-22154.rpm file for all❔

Regardless, the RID in the shared Fx .rpm files for ARM64 should probably be `aarch64` based on my reading of a few Microsoft.DotNet.SharedFramework.Sdk.

Thoughts @nikolam @safern❔

**Comment by dougbu** [comment]
Bit more info: dotnet/runtime may need arch-specific .rpm files because their PlatformManifest.txt files vary between targeting packs e.g.

``` diff
373,374c373,374
< mscordaccore_amd64_amd64_7.0.22.15502.dll|Microsoft.NETCore.App.Ref|7.0.0.0|7.0.22.15502
< mscordaccore_amd64_amd64_7.0.22.15502.dll|Microsoft.NETCore.App.Ref|7.0.0.0|7.0.22.15502
---
> mscordaccore_amd64_arm64_7.0.22.15502.dll|Microsoft.NETCore.App.Ref|7.0.0.0|7.0.22.15502
> mscordaccore_arm64_arm64_7.0.22.15502.dll|Microsoft.NETCore.App.Ref|7.0.0.0|7.0.22.15502
```

We don't have anything similar in dotnet/aspnetcore though the -cm.1 / -cm.2 requirement (for distinct signing) probably still applies. Depending on the upstream expectations and binary content in the installers themselves (❔❔❔), we may just need to fix the duplication in our asset manifests.

**Comment by dougbu** [comment]
Test build matching this iteration: https://dev.azure.com/dnceng/internal/_build/results?buildId=1647192&view=results

**Comment by dougbu** [comment]
To summarize:

1. Is it a bug that our shared Fx RPM is named like aspnetcore-runtime-7.0.0-preview.3.22155.3-**arm64**.rpm❔
2. Do we require arm64-specific targeting pack RPMs❔
3. If (2) is "no", should we remove `--architecture`  from the `fpm` command line when producing the targeting pack RPMs❔
4. I have FR questions open about how to test RPMs on an ARM64 machine. Any suggestions❔

**Comment by NikolaMilosavljevic** [comment]
> To summarize:
> 
> 1. Is it a bug that our shared Fx RPM is named like aspnetcore-runtime-7.0.0-preview.3.22155.3-**arm64**.rpm❔
> 2. Do we require arm64-specific targeting pack RPMs❔
> 3. If (2) is "no", should we remove `--architecture`  from the `fpm` command line when producing the targeting pack RPMs❔
> 4. I have FR questions open about how to test RPMs on an ARM64 machine. Any suggestions❔

1. Package filename does not matter much, but for consistency it would be good if it used the correct arch name `aarch64`.
2. We do need aarch64 targeting pack RPM. Package identity is in package metadata and arch needs to match target platform architecture.

**Comment by dougbu** [comment]
@NikolaMilosavljevic as things stand, I can fix this PR and #40504 but am dubious about getting the 6.0.x fix in before the branch closes today due to component governance problems I'm fixing.

I also won't be able to confirm the new RPMs are usable on a Linux ARM64 machine w/o guidance and an appropriate system. I'm not getting very far in my Teams conversation w/ FR.

**Comment by dougbu** [comment]
Latest test build: https://dev.azure.com/dnceng/internal/_build/results?buildId=1650771&view=results

**Comment by dougbu** [comment]
Now: https://dev.azure.com/dnceng/internal/_build/results?buildId=1650784&view=results

**Comment by dougbu** [comment]
Example names are now
- [aspnetcore-targeting-pack-7.0.0-preview.3.22157.12-cm.1-aarch64.rpm](https://dev.azure.com/dnceng/_apis/resources/Containers/9779258/BlobArtifacts?itemPath=BlobArtifacts%2Faspnetcore-targeting-pack-7.0.0-preview.3.22157.12-cm.1-aarch64.rpm)
- [aspnetcore-targeting-pack-7.0.0-preview.3.22157.12-cm.1-amd64.rpm](https://dev.azure.com/dnceng/_apis/resources/Containers/9779258/BlobArtifacts?itemPath=BlobArtifacts%2Faspnetcore-targeting-pack-7.0.0-preview.3.22157.12-cm.1-amd64.rpm)

1. Should `amd64` be `x64`❔ That would match the RPMs from dotnet/runtime but I don't want to reason from a sample set of one.
2. Should `-cm.1` or `-cm.2` appear after the architecture instead e.g. aspnetcore-targeting-pack-7.0.0-preview.3.22157.12-amd64-cm.1.rpm❔

**Comment by dougbu** [comment]
Fixed `x64` names. New test build is https://dev.azure.com/dnceng/internal/_build/results?buildId=1651818&view=results

**Comment by wtgodbe** [comment]
Validation confirmed the install works - thanks for the help @dougbu and @NikolaMilosavljevic!

**Pull request merged**